src/Shared/Targets/Wrappers/AspNetBufferingTargetWrapper.cs

```diff
@@ -0,0 +1,301 @@
+        /// </remarks>
+        /// <docgen category='Buffering Options' order='100' />
+        [DefaultValue(false)]
+        public bool GrowBufferAsNeeded { get; set; }
+
+        /// <summary>
+        /// Limits the amount of slots that the buffer should grow
+        /// </summary>
+        private int _bufferGrowLimit;
+
+        /// <summary>
+        /// Gets or sets the maximum number of log events that the buffer can keep.
+        /// </summary>
+        /// <docgen category='Buffering Options' order='100' />
+        public int BufferGrowLimit
+        {
+            get => _bufferGrowLimit;
+
+            set
+            {
+                _bufferGrowLimit = value;
+                GrowBufferAsNeeded = (value >= BufferSize);
+            }
+        }
+
+        /// <summary>
+        /// Context for DI
+        /// </summary>
+        private IHttpContextAccessor _httpContextAccessor;
+
+        /// <summary>
+        /// Provides access to the current request HttpContext.
+        /// </summary>
+        /// <returns>HttpContextAccessor or <c>null</c></returns>
+        internal IHttpContextAccessor HttpContextAccessor
+        {
+            get => _httpContextAccessor ?? (_httpContextAccessor = RetrieveHttpContextAccessor());
+            set => _httpContextAccessor = value;
+        }
+
+        private IHttpContextAccessor RetrieveHttpContextAccessor()
+        {
+#if ASP_NET_CORE
+            return ServiceLocator.ResolveService<IHttpContextAccessor>(ResolveService<IServiceProvider>(), LoggingConfiguration);
+#else
+            return new DefaultHttpContextAccessor();
+#endif
+        }
+
+        /// <summary>
+        /// The Key for the buffer dictionary in the HttpContext.Items collection
+        /// </summary>
+        private static readonly object HttpContextItemsKey = new object();
+
+        /// <summary>
+        /// The lock obtained before the dictionary is created or the slot
+        /// in the dictionary is created.
+        /// </summary>
+        private readonly object _lock = new object();
+
+        /// <summary>
+        /// Obtains a slot in the buffer dictionary for 'this' class instance
+        /// If that does not exist, that is created first
+        /// if the dictionary does not exist, that is created first
+        /// </summary>
+        /// <param name="context"></param>
+        /// <returns></returns>
+        private Internal.LogEventInfoBuffer GetOrCreateRequestBuffer(HttpContextBase context) 
+        {
+            // Make sure to create the LogEventInfoBuffer only once in multi-threaded situation.
+            if (context == null)
+            {
+                return null;
+            }
+
+            var bufferDictionary = GetBufferDictionary(context);
+
+            // If the dictionary is missing, create that first
+            if (bufferDictionary == null)
+            {
+                lock (_lock)
```

Comment:
Should we just say lock(context)?

Reply:
See comment above about always creating empty dictionary upfront in middleware, so target-wrapper will only insert itself into the pre-created dictionary on logevent.

Reply:
Completed, see above.